### PR TITLE
add reset method

### PR DESCRIPTION
**Pull request by Stebalien**
Needed for the new stream interface.

This also fixes the tags.

### CODE REVIEW

**Comment by Stebalien** [comment]
@whyrusleeping, @diasdavid

**Comment by diasdavid** [comment]
@Stebalien could you please ensure that interop is still guaranteed? To do that you can:

```
> git clone https://github.com/ipfs/js-ipfs
> cd js-ipfs
> npm install
> cp $GOPATH/bin/ipfs node_modules/go-ipfs-dep/go-ipfs/ipfs
> npm run test:interop
```

**Comment by Stebalien** [comment]
I can't test the reset method (go-ipfs doesn't yet use it) but those tests appear to pass. However, I may not have hacked in the update to go-multiplex correctly as doing so required manually running sed on my local gx repo (I can't just call `gx-go uw` because IPFS won't build in that case).

File: multiplex.go [hunk 1]
```diff
@@ -234,6 +233,10 @@ func (mp *Multiplex) handleIncoming() {
 		mp.chLock.Lock()
 		msch, ok := mp.channels[ch]
 		mp.chLock.Unlock()
+
+		// Why the +1s? 1 is added to the tag if the message is sent by an initiator.
```
Comment (whyrusleeping) [inline multiplex.go]: we do care, its how we avoid collisions in stream IDs

File: multiplex.go [hunk 2]
```diff
@@ -252,7 +255,29 @@ func (mp *Multiplex) handleIncoming() {
 				return
 			}
 
-		case Close:
+		case Reset, Reset + 1:
```
Comment (whyrusleeping) [inline multiplex.go]: Oh, youre not &ing out the flag properly if you have to do this.

File: multiplex.go [hunk 3]
```diff
@@ -252,7 +255,29 @@ func (mp *Multiplex) handleIncoming() {
...
+			if !ok {
+				continue
+			}
+			msch.clLock.Lock()
+
+			// Honestly, this check should never be true... It means we've leaked.
```
Comment (whyrusleeping) [inline multiplex.go]: might be worth logging then

File: multiplex.go [hunk 4]
```diff
@@ -276,23 +301,35 @@ func (mp *Multiplex) handleIncoming() {
 				delete(mp.channels, ch)
 				mp.chLock.Unlock()
 			}
-		default:
+		case Message, Message + 1:
 			if !ok {
 				log.Debugf("message for non-existant stream, dropping data: %d", ch)
+				// Guess initiator status based on the tag.
```
Comment (whyrusleeping) [inline multiplex.go]: this is confusing, we dont need to be guessing anything.

File: multiplex.go (hunk 1 above)
Comment (Stebalien) [inline multiplex.go]: Connection initiator versus stream initiator. We choose stream IDs based on who initiated the connection.

File: multiplex.go (hunk 2 above)
Comment (Stebalien) [inline multiplex.go]: Flag?

File: multiplex.go (hunk 4 above)
Comment (Stebalien) [inline multiplex.go]: We do if we don't recognize the stream and don't know who initiated it.

File: multiplex.go (hunk 2 above)
Comment (Stebalien) [inline multiplex.go]: According to the JS implementation, the tags are actually:

```go
OpenStream
Receiver // (ReceiverMessage)
Initiator // (InitiatorMessage)
ReceiverClose
InitiatorClose
ReceiverReset
InitiatorReset
```

Comment (whyrusleeping) [inline multiplex.go]: . . . . thats kinda dumb. but whatever, carry on

**Comment by Stebalien** [comment]
Here's the JS version of that loop: https://github.com/dignifiedquire/multiplex/blob/master/src/index.js#L300-L320

**Comment by Stebalien** [comment]
Funky? How could you say that! There's a very clear and obvious reason for the receiver/initiator to use different tags. We just haven't thought of it yet...

**Pull request merged**